metrics/src/metrics_manager.rs

```diff
@@ -0,0 +1,135 @@
+    _socket_port: u16,
+    event_channel: ChannelRef<NetworkChannelMsg>,
+    broadcaster: WsSender,
+    connected_clients: Arc<AtomicUsize>,
+    peer_monitors: HashMap<String, PeerMonitor>,
+}
+
+impl MetricsManager {
+    fn name() -> &'static str {
+        "metrics_manager"
+    }
+
+    fn new((event_channel, socket_port): (ChannelRef<NetworkChannelMsg>, u16)) -> Self {
+        let connected_clients = Arc::new(AtomicUsize::new(0));
+        let ws_server = WebSocket::new(WsServer::new(connected_clients.clone()))
+            .expect("Unable to create websocket server");
+        let broadcaster = ws_server.broadcaster();
+
+        Builder::new().name("ws_metrics_server".to_owned()).spawn(move || {
+            let socket = ws_server.bind(("localhost", socket_port))
+                .expect("Unable to bind websocket server");
+            socket.run().expect("Unable to run websocket server");
+        }).expect("Failed to spawn websocket thread");
+        info!("Starting websocket server at port: {}", socket_port);
+
+
+        Self {
+            _socket_port: socket_port,
+            event_channel,
+            broadcaster,
+            connected_clients,
+            peer_monitors: HashMap::new(),
+        }
+    }
+
+    pub fn actor(sys: &impl ActorRefFactory, event_channel: ChannelRef<NetworkChannelMsg>, socket_port: u16) -> Result<MetricsManagerRef, CreateError> {
+        sys.actor_of(
+            Props::new_args(Self::new, (event_channel, socket_port)),
+            Self::name(),
+        )
+    }
+}
+
+impl Actor for MetricsManager {
+    type Msg = MetricsManagerMsg;
+
+    fn pre_start(&mut self, ctx: &Context<Self::Msg>) {
+        // Listen for all network events
+        self.event_channel.tell(Subscribe {
+            actor: Box::new(ctx.myself()),
+            topic: NetworkChannelTopic::NetworkEvents.into(),
+        }, None);
+
+        // Every second, send yourself a message to broadcast the metrics to all connected clients
+        ctx.schedule(Duration::from_secs(1),
+                     Duration::from_secs(1),
+                     ctx.myself(), None,
+                     BroadcastSignal);
+    }
+
+    fn recv(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, sender: Option<BasicActorRef>) {
+        self.receive(ctx, msg, sender);
+    }
+}
+
+impl Receive<BroadcastSignal> for MetricsManager {
+    type Msg = MetricsManagerMsg;
+
+    fn receive(&mut self, _ctx: &Context<Self::Msg>, _msg: BroadcastSignal, _sender: Sender) {
+        use serde_json::to_string;
+        use std::sync::atomic::Ordering::Relaxed;
+
+        if self.connected_clients.load(Relaxed) > 0 {
+            // There are clients connected to the WebSocket
+            let snapshot_data: HashMap<&String, PeerMetrics> = self.peer_monitors.iter_mut()
+                .map(|(k, v)| {
+                    (k, v.snapshot())
+                }).collect();
+            if let Ok(message) = to_string(&snapshot_data) {
```

Comment:
I would prefer to use `serde_json::to_string` here, so it's obvious we are serializing struct to json string.

Reply:
Fixed